power_grid_model_c/power_grid_model/include/power_grid_model/math_solver/newton_raphson_se_solver.hpp

```diff
@@ -170,21 +167,277 @@ template <bool sym> class NewtonRaphsonSESolver {
+        jacobian.dP_dv = -gc_plus_bs;
+        jacobian.dQ_dt = dot(gc_plus_bs, abs_u_other_inv);
+        jacobian.dQ_dv = dot(gs_minus_bc, abs_u_other_inv);
+        return jacobian;
+    }
+
+    void multiply_add_jacobian_blocks(NRSEGainBlock<sym>& block, NRSERhs<sym>& rhs_block, NRSEJacobian const& block_1,
+                                      NRSEJacobian const& block_2, PowerSensorCalcParam<sym> const& power_sensor,
+                                      RealValue<sym> const& calculated_power_p,
+                                      RealValue<sym> const& calculated_power_q) {
+        auto const w_p = diagonal_inverse(power_sensor.p_variance);
+        auto const w_q = diagonal_inverse(power_sensor.q_variance);
+        auto const del_power_p = calculated_power_p - real(power_sensor.value);
+        auto const del_power_q = calculated_power_q - imag(power_sensor.value);
+
+        // matrix multiplication of F_k^T . w_k . F_k
+        block.g_P_theta() += dot(w_p, block_1.dP_dt, block_2.dP_dt) + dot(w_q, block_1.dQ_dt, block_2.dQ_dt);
+        block.g_P_v() += dot(w_p, block_1.dP_dv, block_2.dP_dv) + dot(w_q, block_1.dQ_dv, block_2.dQ_dv);
+        block.g_Q_theta() += dot(w_p, block_1.dP_dt, block_2.dP_dv) + dot(w_q, block_1.dQ_dt, block_2.dQ_dv);
+        block.g_Q_v() += dot(w_p, block_1.dP_dv, block_2.dP_dt) + dot(w_q, block_1.dQ_dv, block_2.dQ_dt);
+
+        // matrix multiplication of F_k^T . w_k . (z - f(x))
+        rhs_block.eta_theta() += dot(w_p, block_1.dP_dt, del_power_p) + dot(w_q, block_1.dQ_dt, del_power_q);
+        rhs_block.eta_v() += dot(w_p, block_1.dP_dv, del_power_p) + dot(w_q, block_1.dQ_dv, del_power_q);
+    }
+
+    void add_injection_jacobian(NRSEGainBlock<sym>& block, NRSEJacobian const& jacobian_block) {
+        block.q_P_theta() += jacobian_block.dP_dt;
+        block.q_P_v() += jacobian_block.dP_dv;
+        block.q_Q_theta() += jacobian_block.dQ_dt;
+        block.q_Q_v() += jacobian_block.dQ_dv;
+    }
+
+    auto ui_uj_cos_ij(ComplexValue<sym> const& ui, ComplexValue<sym> const& uj) {
+        return vector_outer_product(real(ui), real(uj)) + vector_outer_product(imag(ui), imag(uj));
+    }
+
+    auto ui_uj_sin_ij(ComplexValue<sym> const& ui, ComplexValue<sym> const& uj) {
+        return vector_outer_product(imag(ui), real(uj)) - vector_outer_product(real(ui), imag(uj));
+    }
+
+    auto g_sin_minus_b_cos(ComplexTensor<sym> const& yij, ComplexValue<sym> const& ui, ComplexValue<sym> const& uj) {
+        return real(yij) * ui_uj_sin_ij(ui, uj) - imag(yij) * ui_uj_cos_ij(ui, uj);
+    }
+
+    auto g_cos_plus_b_sin(ComplexTensor<sym> const& yij, ComplexValue<sym> const& ui, ComplexValue<sym> const& uj) {
+        return real(yij) * ui_uj_cos_ij(ui, uj) + imag(yij) * ui_uj_sin_ij(ui, uj);
+    }
+
     double iterate_unknown(ComplexValueVector<sym>& u) {
         double max_dev = 0.0;
 
         for (Idx bus = 0; bus != n_bus_; ++bus) {
-            // angle
+            // accumulate unknown
```

Comment:
"unknown" sounds weird, especially when talking about "accumulate unknown". is there another name you can use for this? or maybe make it plural: "unknowns" vs "accumulate unknowns". That actually has a mathematical meaning

Reply:
Changed to `accumulate the unknown variable`